site_tools/image_deployer.py

```diff
@@ -0,0 +1,297 @@
+        for i in connection.image.images(
+            visibility="public"
+        )
+    ]
+    return images
+
+
+def should_sync_image(image_disk_name, site_images, current):
+    if image_disk_name in site_images:
+        logging.info(f"Image {image_disk_name} already in site images.")
+        image = image_connection.image.find_image(image_disk_name)
+        image_properties = image.properties
+        logging.debug(f"Image properties: {image_properties}")
+        image_current_value = image_properties.get("current", None)
+        logging.debug(f"Image {image_disk_name} current value: {image_current_value}")
+        if image_current_value is not None and image_current_value == current:
+            logging.info(f"Image {image_disk_name} is already current.")
+            return False
+    return True
+
+
+def download_object_to_file(storage_url, path, file_name):
+    url = f"{storage_url}/{path}/{file_name}"
+    response = requests.get(url, stream=True)
+
+    if response.status_code != 200:
+        raise Exception(f"Error downloading object {file_name}: {response.content}")
+
+    with tempfile.NamedTemporaryFile(delete=False) as temp_file:
+        for chunk in response.iter_content(chunk_size=65536):
+            temp_file.write(chunk)
+            temp_file.flush()
+    logging.debug(f"Downloaded object to {temp_file.name}.")
+    return temp_file
+
+
+def upload_image_to_glance(image_connection,
+                           image_prefix,
+                           image_disk_name,
+                           image_file_name,
+                           disk_format,
+                           manifest_data):
+    image_prefix_name = image_prefix + image_disk_name
+
+    # TODO: need admin creds to make visibility public instead of private
+    # right now I am using app creds that just have a member role
+    logging.info(f"Uploading image {image_disk_name} to Glance.")
+    with open(image_file_name, "rb") as image_data:
+        new_image = image_connection.create_image(name=image_prefix_name,
+                                                  disk_format=disk_format,
+                                                  container_format="bare",
+                                                  visibility="private",
+                                                  data=image_data,
+                                                  **manifest_data)
+    logging.info(f"Uploaded image {new_image.name}.")
+    return new_image
+
+
+def archive_image(image_connection, image_disk_name, new_image):
+    # TODO: if things go sideways in here we may be in a bad state, add more
+    # error handling.
+    existing_images = list(image_connection.image.images(name=image_disk_name))
+    # TODO: rename the old image: this should only ever be 1. assert that instead?
+    if len(existing_images) > 0:
+        logging.info(f"Renaming existing image {image_disk_name}.")
+        archive_date = datetime.datetime.now().strftime("%Y%m%d_%H%M")
```

Comment:
for the archive date, we should populate this from the image's metadata, probably the build-date?